README.md
# serde-RESP
Redis RESP protocol serialization and deserialization with serde.
[RESP Specification](https://redis.io/topics/protocol)

To use the crate, add this to your `Cargo.toml`:
```
[dependencies]
serde-resp = "0.3.1"
```

## Usage
**IMPORTANT: Do NOT (de)serialize with any other types besides `RESP/RESPType`! You may get panic or incorrect results!**
[Why?](#why-resptype-wrapper)

Here are the Redis RESP types and their corresponding Rust types for (de)serialize operations.

- `Simple String`
    + `RESP::SimpleString(String)`
- `Error`
    + `RESP::Error(String)`
- `Integer`
    + `RESP::Integer(i64)`
- `Bulk String`
    + `RESP::BulkString(Option<Vec<u8>>)`
        + Use `None` for null bulk strings and `Some` for non-null ones.
- `Array`
    + `RESP::Array(Option<Vec<RESP>>)`
        + Use `None` for null arrays and `Some` for non-null ones.

To serialize, use [ser::to_string](https://docs.rs/serde_resp/0.3.1/serde_resp/ser/fn.to_string.html)
or [ser::to_writer](https://docs.rs/serde_resp/0.3.1/serde_resp/ser/fn.to_writer.html).

To deserialize, use [de::from_str](https://docs.rs/serde_resp/0.3.1/serde_resp/de/fn.from_str.html) 
or [de::from_reader](https://docs.rs/serde_resp/0.3.1/serde_resp/de/fn.from_reader.html)
or [de::from_buf_reader](https://docs.rs/serde_resp/0.3.1/serde_resp/de/fn.from_buf_reader.html).

For usage examples, refer to https://docs.rs/serde_resp/0.3.1/serde_resp/enum.RESPType.html

## Macros
Since 0.3.0, you can start using very handy macros! Here is a demo:

```rust
use serde_resp::{array, array_null, bulk, bulk_null, de, err_str, int, ser, simple, RESP};
let resp_array = array![
    simple!("simple string".to_owned()),
    err_str!("error string".to_owned()),
    int!(42),
    bulk!(b"bulk string".to_vec()),
    bulk_null!(),
    array![
        simple!("arrays of arrays!".to_owned()),
        array![simple!("OK ENOUGH!".to_owned())],
    ],
    array_null!(),
];
let serialized = ser::to_string(&resp_array).unwrap();
assert_eq!(
    "*7\r\n+simple string\r\n-error string\r\n:42\r\n$11\r\nbulk string\r\n$-1\r\n*2\r\n+arrays of arrays!\r\n*1\r\n+OK ENOUGH!\r\n*-1\r\n",
    serialized
);
let deserialized = de::from_str(&serialized).unwrap();
assert_eq!(resp_array, deserialized);
```

## Documentation

https://docs.rs/serde_resp/0.3.1/serde_resp

## Why RESPType Wrapper?
The RESP protocol does not fit the serde data model well. 
- RESP has no structured types (besides array).
  This means (de)serializing structured data like `struct`, `enum`, or `tuple` does not make sense.
- RESP has multiple string types (simple and error string).
  So what should Rust `String` map to? Simple or error string? Using what serde offers here will introduce a lot of ambiguity.
  Another problem is that serde assumes there is only one string type so I need to apply some "hacks" to make (de)serializer work correctly.
  These "hacks" involve (de)serialize methods that users are not supposed to use.
- RESP has multiple null types (null bulk string and null array). Similar to above, a lot of ambiguity and "hacks".

A bonus reason is this makes it easier to define array, which can be recursive.

I know this is very strict and tedious to write. But less ambiguity is always a good thing.
Plus now with macros, it should be more pleasant to use now.

CHANGELOG.md
## 2021-02-16, Version v0.3.1
### Commits
- [[`fa9d1da463`](https://github.com/DEDZTBH/serde-RESP/commit/fa9d1da463421cff4b82a1efb4681c6ab49e779d)] v0.3.1 (DEDZTBH)
- [[`8d1ac8f318`](https://github.com/DEDZTBH/serde-RESP/commit/8d1ac8f318153d16abe8cd9f58b3d396971fb779)] changelog update (DEDZTBH)

### Stats
```diff
 CHANGELOG.md  | 18 ++++++++++++++++++
 Cargo.toml    |  2 +-
 README.md     | 41 ++++++++++++++++++++++++++++++-----------
 src/de.rs     |  9 ++++++---
 src/error.rs  |  8 +++++---
 src/lib.rs    | 14 +++++++++-----
 src/macros.rs | 17 +++++++++--------
 7 files changed, 78 insertions(+), 31 deletions(-)
```


## 2021-02-15, Version v0.3.0
### Commits
- [[`07a18496e3`](https://github.com/DEDZTBH/serde-RESP/commit/07a18496e32cbb4ab55456a87ad726db0cb8514d)] v3.0 (DEDZTBH)
- [[`afb8c76b68`](https://github.com/DEDZTBH/serde-RESP/commit/afb8c76b68a082e2dfe68f682b5f3c54b66d8e3f)] update readme (DEDZTBH)
- [[`a818052e11`](https://github.com/DEDZTBH/serde-RESP/commit/a818052e119740188fe506594dd247621162871f)] update changelog (DEDZTBH)

### Stats
```diff
 CHANGELOG.md  |  14 ++++-
 Cargo.toml    |   2 +-
 README.md     |  61 ++++++++++++-----
 src/de.rs     |   8 +-
 src/lib.rs    | 216 +++++++++++++++++++++++++++++++++--------------------------
 src/macros.rs |  56 +++++++++++++++-
 6 files changed, 243 insertions(+), 114 deletions(-)
```


## 2021-02-15, Version v0.2.0
### Commits
- [[`8ecef61f09`](https://github.com/DEDZTBH/serde-RESP/commit/8ecef61f09ec807aefbb1590765e4ec378b0ed28)] v0.2.0 (DEDZTBH)
- [[`08cd73c14e`](https://github.com/DEDZTBH/serde-RESP/commit/08cd73c14eaafc5dabbff6bd2d34b1730adfa0fe)] removed license (DEDZTBH)

### Stats
```diff
 Cargo.toml |  3 +--
 src/de.rs  | 50 ++++++++++++++++++++++++++++++++++++++------------
 src/lib.rs | 38 ++++++++++----------------------------
 3 files changed, 49 insertions(+), 42 deletions(-)
```




tests/tests.rs
use serde_resp::ser;
use serde_resp::RESPType;

#[test]
fn basic_test() {
    let obj = RESPType::SimpleString("magic".to_owned());
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("+magic\r\n".to_owned(), serialized);
}


tests/ser_tests.rs
use serde_resp::{ser, RESPType};

#[test]
fn serialize_simple_string() {
    let obj = RESPType::SimpleString("OK".to_owned());
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("+OK\r\n".to_owned(), serialized);
}

#[test]
fn serialize_error() {
    let obj = RESPType::Error("ERR unknown command 'foobar'".to_owned());
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("-ERR unknown command 'foobar'\r\n".to_owned(), serialized);
    let obj = RESPType::Error(
        "WRONGTYPE Operation against a key holding the wrong kind of value".to_owned(),
    );
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!(
        "-WRONGTYPE Operation against a key holding the wrong kind of value\r\n".to_owned(),
        serialized
    );
}

#[test]
fn serialize_integer() {
    let obj = RESPType::Integer(1000);
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!(":1000\r\n".to_owned(), serialized);
}

#[test]
fn serialize_bulk_string() {
    // Regular Example
    let obj = RESPType::BulkString(Some(b"foobar"));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("$6\r\nfoobar\r\n".to_owned(), serialized);

    // Empty
    let obj = RESPType::BulkString(Some(b""));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("$0\r\n\r\n".to_owned(), serialized);

    // Null
    let obj = RESPType::BulkString(None);
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("$-1\r\n".to_owned(), serialized);
}

#[test]
fn serialize_array() {
    let obj = RESPType::Array(Some(&[]));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("*0\r\n".to_owned(), serialized);

    // Regular Example
    let obj = RESPType::Array(Some(&[
        RESPType::BulkString(Some(b"foo")),
        RESPType::BulkString(Some(b"bar")),
    ]));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("*2\r\n$3\r\nfoo\r\n$3\r\nbar\r\n".to_owned(), serialized);

    // Another Regular Example
    let obj = RESPType::Array(Some(&[
        RESPType::Integer(1),
        RESPType::Integer(2),
        RESPType::Integer(3),
    ]));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("*3\r\n:1\r\n:2\r\n:3\r\n".to_owned(), serialized);

    // Mixed Type
    let obj = RESPType::Array(Some(&[
        RESPType::Integer(1),
        RESPType::Integer(2),
        RESPType::Integer(3),
        RESPType::Integer(4),
        RESPType::BulkString(Some(b"foobar")),
    ]));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!(
        "*5\r\n:1\r\n:2\r\n:3\r\n:4\r\n$6\r\nfoobar\r\n".to_owned(),
        serialized
    );

    // Null Array
    let obj = RESPType::Array(None);
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!("*-1\r\n".to_owned(), serialized);

    // Arrays of Arrays
    let subarr1 = [
        RESPType::Integer(1),
        RESPType::Integer(2),
        RESPType::Integer(3),
    ];
    let subarr2 = [
        RESPType::SimpleString("Foo".to_owned()),
        RESPType::Error("Bar".to_owned()),
    ];
    let arr = [
        RESPType::Array(Some(&subarr1)),
        RESPType::Array(Some(&subarr2)),
    ];
    let obj = RESPType::Array(Some(&arr));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!(
        "*2\r\n*3\r\n:1\r\n:2\r\n:3\r\n*2\r\n+Foo\r\n-Bar\r\n".to_owned(),
        serialized
    );

    // Null elements in Arrays
    let arr = [
        RESPType::BulkString(Some(b"foo")),
        RESPType::BulkString(None),
        RESPType::BulkString(Some(b"bar")),
    ];
    let obj = RESPType::Array(Some(&arr));
    let serialized = ser::to_string(&obj).unwrap();
    assert_eq!(
        "*3\r\n$3\r\nfoo\r\n$-1\r\n$3\r\nbar\r\n".to_owned(),
        serialized
    );
}


src/de.rs
use crate::{Error, RESPType, Result};

use serde::de::{self, DeserializeOwned, DeserializeSeed, SeqAccess, Visitor};
use serde::Deserialize;

use std::fmt;
use std::io::{BufRead, BufReader, Cursor, Read};

/// Deserializer for RESP format
pub struct Deserializer<'de, R: BufRead> {
    reader: &'de mut R,
}

impl<'de, R: BufRead> Deserializer<'de, R> {
    /// Method for building Deserializer
    pub fn from_buf_reader(reader: &'de mut R) -> Deserializer<'de, R> {
        Deserializer { reader }
    }
}

/// Deserialize from str.
///
/// This function simple wraps the `&str` with `Cursor` and calls [from_buf_reader](from_buf_reader).
///
/// # Errors
/// Please refer to [Error](Error)
pub fn from_str<T>(s: &str) -> Result<T>
where
    T: DeserializeOwned,
{
    let mut reader = Cursor::new(s);
    from_buf_reader(&mut reader)
}

/// Deserialize from reader with `Read` trait.
///
/// This function simply wraps the reader with a `BufReader` and calls [from_buf_reader](from_buf_reader).
/// If your reader has `BufRead` trait, use [from_buf_reader](from_buf_reader) instead.
///
/// # Errors
/// Please refer to [Error](Error)
pub fn from_reader<T, R>(reader: &mut R) -> Result<T>
where
    T: DeserializeOwned,
    R: Read,
{
    let mut reader = BufReader::new(reader);
    from_buf_reader(&mut reader)
}

/// Deserialize from reader with `BufRead` trait.
///
/// # Errors
/// Please refer to [Error](Error)
pub fn from_buf_reader<T, R>(reader: &mut R) -> Result<T>
where
    T: DeserializeOwned,
    R: BufRead,
{
    let mut deserializer = Deserializer::from_buf_reader(reader);
    let t = T::deserialize(&mut deserializer)?;
    Ok(t)
}

impl<'de, R: BufRead> Deserializer<'de, R> {
    // read until LF, trim end, and parse to isize.
    fn read_isize(&mut self) -> Result<isize> {
        let mut buffer = String::new();
        self.reader.read_line(&mut buffer)?;
        let trimmed = buffer.trim_end();
        match trimmed.parse::<isize>() {
            Ok(x) => Ok(x),
            Err(_) => Err(Error::Syntax),
        }
    }
}

impl<'de, 'a, R: BufRead> de::Deserializer<'de> for &'a mut Deserializer<'de, R> {
    type Error = Error;

    // You see, this is a bit hacky...
    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let mut buf = [0u8; 1];
        self.reader.read_exact(&mut buf)?;
        match buf[0] {
            b'+' => self.deserialize_str(visitor),      // SimpleString
            b'-' => self.deserialize_string(visitor),   // Error
            b':' => self.deserialize_i64(visitor),      // Integer
            b'$' => self.deserialize_byte_buf(visitor), // BulkString
            b'*' => self.deserialize_seq(visitor),      // Array
            _ => return Err(Error::Syntax),
        }
    }

    fn deserialize_bool<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_i8<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_i16<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_i32<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_i64<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let mut buffer = String::new();
        self.reader.read_line(&mut buffer)?;
        match buffer.trim_end().parse::<i64>() {
            Ok(x) => visitor.visit_i64(x),
            Err(_) => Err(Error::Syntax),
        }
    }

    fn deserialize_u8<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_u16<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_u32<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_u64<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_f32<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_f64<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_char<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    // SimpleString
    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let mut buffer = String::new();
        self.reader.read_line(&mut buffer)?;
        visitor.visit_str(buffer.trim_end())
    }

    // Error
    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let mut buffer = String::new();
        self.reader.read_line(&mut buffer)?;
        visitor.visit_string(buffer.trim_end().to_string())
    }

    fn deserialize_bytes<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    // BulkString
    fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let x = self.read_isize()?;
        if x < 0 {
            return visitor.visit_none();
        }
        let mut buffer = vec![0u8; (x + 2) as usize]; // also read CRLF
        self.reader.read_exact(&mut buffer)?;
        if buffer.split_off(x as usize) != b"\r\n" {
            return Err(Error::Syntax); // Not CRLF
        }
        visitor.visit_byte_buf(buffer)
    }

    fn deserialize_option<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_unit<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_unit_struct<V>(self, _name: &'static str, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_newtype_struct<V>(self, _name: &'static str, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    // Deserialization of compound types like sequences and maps happens by
    // passing the visitor an "Access" object that gives it the ability to
    // iterate through the data contained in the sequence.
    fn deserialize_seq<V>(mut self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let x = self.read_isize()?;
        if x < 0 {
            return visitor.visit_unit();
        }
        visitor.visit_seq(RESPArray::new(&mut self, x as usize))
    }

    fn deserialize_tuple<V>(self, _len: usize, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_tuple_struct<V>(
        self,
        _name: &'static str,
        _len: usize,
        _visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_map<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_struct<V>(
        self,
        _name: &'static str,
        _fields: &'static [&'static str],
        _visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_enum<V>(
        self,
        _name: &'static str,
        _variants: &'static [&'static str],
        _visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_identifier<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_ignored_any<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }
}

struct RESPArray<'a, 'de: 'a, R: BufRead> {
    de: &'a mut Deserializer<'de, R>,
    remain_len: usize,
}

impl<'a, 'de, R: BufRead> RESPArray<'a, 'de, R> {
    fn new(de: &'a mut Deserializer<'de, R>, len: usize) -> Self {
        RESPArray {
            de,
            remain_len: len,
        }
    }
}

// `SeqAccess` is provided to the `Visitor` to give it the ability to iterate
// through elements of the sequence.
impl<'de, 'a, R: BufRead> SeqAccess<'de> for RESPArray<'a, 'de, R> {
    type Error = Error;

    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>>
    where
        T: DeserializeSeed<'de>,
    {
        if self.remain_len == 0 {
            return Ok(None);
        }
        self.remain_len -= 1;
        seed.deserialize(&mut *self.de).map(Some)
    }

    fn size_hint(&self) -> Option<usize> {
        Some(self.remain_len)
    }
}

struct RESPTypeVisitor;

impl<'de> Visitor<'de> for RESPTypeVisitor {
    type Value = RESPType;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("A RESP value")
    }

    fn visit_i64<E>(self, v: i64) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::Integer(v))
    }

    // SimpleString
    fn visit_str<E>(self, v: &str) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::SimpleString(v.to_string()))
    }

    // Error
    fn visit_string<E>(self, v: String) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::Error(v))
    }

    // BulkString
    fn visit_byte_buf<E>(self, v: Vec<u8>) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::BulkString(Some(v)))
    }

    // null BulkString
    fn visit_none<E>(self) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::BulkString(None))
    }

    // null Array
    fn visit_unit<E>(self) -> std::result::Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(RESPType::Array(None))
    }

    fn visit_seq<A>(
        self,
        mut seq: A,
    ) -> std::result::Result<Self::Value, <A as SeqAccess<'de>>::Error>
    where
        A: SeqAccess<'de>,
    {
        let mut arr: Vec<RESPType> = Vec::with_capacity(seq.size_hint().unwrap_or_default());
        loop {
            match seq.next_element()? {
                None => break,
                Some(elem) => arr.push(elem),
            };
        }
        Ok(RESPType::Array(Some(arr)))
    }
}

impl<'de> Deserialize<'de> for RESPType {
    fn deserialize<D>(
        deserializer: D,
    ) -> std::result::Result<Self, <D as de::Deserializer<'de>>::Error>
    where
        D: de::Deserializer<'de>,
    {
        deserializer.deserialize_any(RESPTypeVisitor)
    }
}


src/lib.rs
//! # serde-RESP
//! Redis RESP protocol serialization and deserialization with serde.
//! [Read Specification](https://redis.io/topics/protocol)
//!
//! ## Usage
//! **IMPORTANT: Do NOT (de)serialize with any other types besides `RESP/RESPType`! You may get panic or incorrect results!**
//! [Why?](https://github.com/DEDZTBH/serde-RESP/blob/master/README.md#why-resptype-wrapper)
//!
//! Here are the RESP types and their corresponding Rust types for (de)serialize operations.
//!
//! - `Simple String`
//!     + [RESP::SimpleString(String)](RESPType::SimpleString)
//! - `Error`
//!     + [RESP::Error(String)](RESPType::Error)
//! - `Integer`
//!     + [RESP::Integer(i64)](RESPType::Integer)
//! - `Bulk String`
//!     + [RESP::BulkString(Option<Vec<u8>>)](RESPType::BulkString)
//!         + Use `None` for null bulk strings and `Some` for non-null ones.
//! - `Array`
//!     + [RESP::Array(Option<Vec<RESP>>)](RESPType::Array)
//!         + Use `None` for null arrays and `Some` for non-null ones.
//!
//! To serialize, use [ser::to_string](ser::to_string) or [ser::to_writer](ser::to_writer).
//!
//! To deserialize, use [de::from_str](de::from_str) or [de::from_reader](de::from_reader) or [de::from_buf_reader](de::from_buf_reader).
//!
//! For usage examples, refer to [RESP](RESP)
//!
//! ## Macros
//!
//! Since 0.3.0, you can start using very handy macros! Here is a demo:
//! ```
//!     use serde_resp::{array, array_null, bulk, bulk_null, de, err_str, int, ser, simple, RESP};
//!     let resp_array = array![
//!         simple!("simple string".to_owned()),
//!         err_str!("error string".to_owned()),
//!         int!(42),
//!         bulk!(b"bulk string".to_vec()),
//!         bulk_null!(),
//!         array![
//!             simple!("arrays of arrays!".to_owned()),
//!             array![simple!("OK ENOUGH!".to_owned())],
//!         ],
//!         array_null!(),
//!     ];
//!     let serialized = ser::to_string(&resp_array).unwrap();
//!     assert_eq!(
//!         "*7\r\n+simple string\r\n-error string\r\n:42\r\n$11\r\nbulk string\r\n$-1\r\n*2\r\n+arrays of arrays!\r\n*1\r\n+OK ENOUGH!\r\n*-1\r\n",
//!         serialized
//!     );
//!     let deserialized = de::from_str(&serialized).unwrap();
//!     assert_eq!(resp_array, deserialized);
//! ```

pub mod de;
mod error;
mod macros;
pub mod ser;

pub use error::{Error, Result};

/// This enum creates a one-to-one type mapping with RESP types.
/// Please only use variants of this type for (de)serialize operations.
#[derive(Eq, PartialEq, Clone, Debug)]
pub enum RESPType {
    /// Correspond to simple string in RESP.
    /// Also refer to [simple!](simple!) macro.
    ///
    /// # Examples
    /// ```
    /// use serde_resp::{de, ser, simple, RESP};
    ///
    /// /// Serialization
    /// let obj = simple!("OK".to_owned()); // equivalent to RESP::SimpleString("OK".to_owned());
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("+OK\r\n".to_owned(), serialized);
    ///
    /// /// Deserialization
    /// let deserialized: RESP = de::from_str("+OK\r\n").unwrap();
    /// assert_eq!(simple!("OK".to_owned()), deserialized);
    /// ```
    SimpleString(String),
    /// Correspond to error string in RESP.
    /// Also refer to [err_str!](err_str!) macro.
    ///
    /// # Examples
    /// ```
    /// use serde_resp::{de, err_str, ser, RESP};
    ///
    /// /// Serialization
    /// // Example 1
    /// let obj = err_str!("ERR unknown command 'foobar'".to_owned()); // equivalent to RESP::Error(...)
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("-ERR unknown command 'foobar'\r\n".to_owned(), serialized);
    ///
    /// // Example 2
    /// let obj = err_str!("WRONGTYPE Operation against a key holding the wrong kind of value".to_owned());
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!(
    ///     "-WRONGTYPE Operation against a key holding the wrong kind of value\r\n".to_owned(),
    ///     serialized
    /// );
    ///
    /// /// Deserialization
    /// let deserialized: RESP = de::from_str("-ERR unknown command 'foobar'\r\n").unwrap();
    /// assert_eq!(err_str!("ERR unknown command 'foobar'".to_owned()), deserialized);
    /// ```
    Error(String),
    /// Correspond to integer in RESP.
    /// Also refer to [int!](int!) macro.
    ///
    /// # Examples
    /// ```
    /// use serde_resp::{de, int, ser, RESP};
    ///
    /// /// Serialization
    /// // Regular Example
    /// let obj = int!(1000); // equivalent to RESP::Integer(1000);
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!(":1000\r\n".to_owned(), serialized);
    ///
    /// /// Deserialization
    /// let deserialized: RESP = de::from_str(":1000\r\n").unwrap();
    /// assert_eq!(int!(1000), deserialized);
    /// ```
    Integer(i64),
    /// Correspond to bulk string in RESP. Use `None` for null bulk string and Some for non-null ones.
    /// Also refer to [bulk!](bulk!) macro and [bulk_null!](bulk_null!) macro.
    ///
    /// According to specification, bulk string is binary-safe so it is NOT recommended to use [ser::to_string](ser::to_string) (may cause [Error::FromUtf8](Error::FromUtf8)).
    /// Use [ser::to_writer](ser::to_writer) to write to a byte buffer instead.
    ///
    /// # Examples
    /// ```
    /// use serde_resp::{bulk, bulk_null, de, ser, RESP};
    ///
    /// /// Serialization
    /// // Regular Example
    /// let obj = bulk!(b"foobar".to_vec()); // equivalent to RESP::BulkString(Some(...))
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("$6\r\nfoobar\r\n".to_owned(), serialized);
    ///
    /// // Empty
    /// let obj = bulk!(b"".to_vec());
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("$0\r\n\r\n".to_owned(), serialized);
    ///
    /// // Null
    /// let obj = bulk_null!(); // equivalent to RESP::BulkString(None)
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("$-1\r\n".to_owned(), serialized);
    ///
    /// /// Deserialization
    /// // Regular Example
    /// let deserialized: RESP = de::from_str("$6\r\nfoobar\r\n").unwrap();
    /// assert_eq!(bulk!(b"foobar".to_vec()), deserialized);
    ///
    /// // Empty
    /// let deserialized: RESP = de::from_str("$0\r\n\r\n").unwrap();
    /// assert_eq!(bulk!(b"".to_vec()), deserialized);
    ///
    /// // Null
    /// let deserialized: RESP = de::from_str("$-1\r\n").unwrap();
    /// assert_eq!(bulk_null!(), deserialized);
    /// ```
    BulkString(Option<Vec<u8>>),
    /// Correspond to array in RESP. Use None for null array and Some for non-null ones.
    /// Also refer to [array!](array!) macro and [array_null!](array_null!) macro.
    ///
    /// Mixed type, arrays of arrays are allowed.
    ///
    /// # Examples
    /// ```
    /// use serde_resp::{array, array_null, bulk, bulk_null, de, err_str, int, ser, simple, RESP};
    ///
    /// /// Serialization
    /// // Empty
    /// let obj = array![]; // equivalent to RESP::Array(Some(vec![]))
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("*0\r\n".to_owned(), serialized);
    ///
    /// // Regular Example
    /// let obj = array![
    ///     bulk!(b"foo".to_vec()),
    ///     bulk!(b"bar".to_vec()),
    /// ]; // equivalent to RESP::Array(Some(vec![bulk!(...), bulk!(...)]))
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("*2\r\n$3\r\nfoo\r\n$3\r\nbar\r\n".to_owned(), serialized);
    ///
    /// // Another Regular Example
    /// let obj = array![
    ///     int!(1),
    ///     int!(2),
    ///     int!(3),
    /// ];
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("*3\r\n:1\r\n:2\r\n:3\r\n".to_owned(), serialized);
    ///
    /// // Mixed Type
    /// let obj = array![
    ///     int!(1),
    ///     int!(2),
    ///     int!(3),
    ///     int!(4),
    ///     bulk!(b"foobar".to_vec()),
    /// ];
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!(
    ///     "*5\r\n:1\r\n:2\r\n:3\r\n:4\r\n$6\r\nfoobar\r\n".to_owned(),
    ///     serialized
    /// );
    ///
    /// // Null Array
    /// let obj = array_null!(); // equivalent to RESP::Array(None)
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!("*-1\r\n".to_owned(), serialized);
    ///
    /// // Arrays of Arrays
    /// let obj = array![
    ///     array![
    ///         int!(1),
    ///         int!(2),
    ///         int!(3),
    ///     ],
    ///     array![
    ///         simple!("Foo".to_owned()),
    ///         err_str!("Bar".to_owned()),
    ///     ],
    /// ];
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!(
    ///     "*2\r\n*3\r\n:1\r\n:2\r\n:3\r\n*2\r\n+Foo\r\n-Bar\r\n".to_owned(),
    ///     serialized
    /// );
    ///
    /// // Null elements in Arrays
    /// let obj = array![
    ///     bulk!(b"foo".to_vec()),
    ///     bulk_null!(),
    ///     bulk!(b"bar".to_vec()),
    /// ];
    /// let serialized = ser::to_string(&obj).unwrap();
    /// assert_eq!(
    ///     "*3\r\n$3\r\nfoo\r\n$-1\r\n$3\r\nbar\r\n".to_owned(),
    ///     serialized
    /// );
    ///
    /// /// Deserialization
    /// // Null
    /// let deserialized: RESP = de::from_str("*-1\r\n").unwrap();
    /// assert_eq!(array_null!(), deserialized);
    ///
    /// // Mixed Type
    /// let deserialized: RESP = de::from_str("*2\r\n:1\r\n$6\r\nfoobar\r\n").unwrap();
    /// let expected = array![
    ///     int!(1),
    ///     bulk!(b"foobar".to_vec()),
    /// ];
    /// assert_eq!(expected, deserialized);
    ///
    /// // Arrays of Arrays with Null Bulk String
    /// let deserialized: RESP = de::from_str("*3\r\n*3\r\n:1\r\n:2\r\n:3\r\n$-1\r\n*2\r\n+Foo\r\n-Bar\r\n").unwrap();
    /// let expected = array![
    ///     int!(1),
    ///     bulk!(b"foobar".to_vec()),
    /// ];
    /// let expected = array![
    ///     array![
    ///         int!(1),
    ///         int!(2),
    ///         int!(3),
    ///     ],
    ///     bulk_null!(),
    ///     array![
    ///         simple!("Foo".to_owned()),
    ///         err_str!("Bar".to_owned()),
    ///     ],
    /// ];
    /// assert_eq!(expected, deserialized);
    /// ```
    Array(Option<Vec<RESPType>>),
}

/// Refer to [RESPType](RESPType). This is just an alias.
pub type RESP = RESPType;
